### ISSUE TITLE
file names in source maps are set to all lowercase

### ISSUE THREAD

**Issue by benjaminRomano** [ivogabe/gulp-typescript#255]
The file names in the source map's source property are being set to all lowercase.

For example, `['exampleFileName.ts']` is being changed to `['examplefilename.ts']`.

I'm currently using Visual Studio Code's debugger and my source maps aren't work because the debugger expects the file names to have the correct casing.

I was able to fix the problem locally by removing the call to toLowerCase in the normalizePath function in utils.ts. Is there a reason why the file names are lowercase or would it be fine to change that?

**Comment by ivogabe** [comment]
Thanks for reporting, I will take a look at this. The `toLowerCase` is necessary, for case insensitive file systems. The result of normalizePath should not be propagated to the source map though.

**Comment by mtebenev** [comment]
I confirm the issue. Also it's strange that on some modules the names are ok (i.e. not lowered).

**Comment by benjaminRomano** [comment]
Any update on this? This is a big problem when using visual studio code.

**Comment by pyrho** [comment]
I confirm the issue, prevents from debugging with VSCode.
I also confirm the workaround proposed by @benjaminRomano .

> The toLowerCase is necessary, for case insensitive file systems
Not sure I understand why this is required for case insensitive FS, since they are.... case insensitive.
Maybe we could just get rid of that bit entirely.

**Comment by ivogabe** [comment]
> Not sure I understand why this is required for case insensitive FS, since they are.... case insensitive.
Maybe we could just get rid of that bit entirely.

If the compiler has already seen `foo.ts`, and `Foo.ts` is referenced from some other file, it should not try to fetch `Foo.ts` (which would resolve to the same content of `foo.ts`). The same file would be compiled twice this way, which can give errors that an identifier has already been imported. This normalized path should only be used internal and not be exposed however.

**Comment by pyrho** [comment]
Right, thanks for the reply.
Any pointers on how this should be fixed ? Maybe I can come up with a PR.

**Comment by ivogabe** [comment]
I'm not entirely sure, but it's probably `lib/output.ts` or `lib/compiler.ts`.

**Comment by mbakturk** [comment]
I encountered same problem, all files are mapped as lower case name

**Comment by holdfenytolvaj** [comment]
> If the compiler has already seen foo.ts, and Foo.ts is referenced from some other file, it should not try to fetch Foo.ts
Not sure whether it is a good feature, since if somebody deploy his/her code from mac/win to linux server, it can cause issues. It is better to detect issues as early as possible, would be preferable to throw an error.

**Comment by ivogabe** [comment]
@holdfenytolvaj What I described matches the behavior of `tsc`.

**Comment by holdfenytolvaj** [comment]
@ivogabe strange, would be interesting to know why... Since it is hiding errors, that could be revealed at compile time instead of runtime. One of the biggest advertisement of js the code portability is hurt by this.

**Comment by ivogabe** [comment]
@holdfenytolvaj You can use `forceConsistentCasingInFileNames` to catch those issues. See the [wiki](https://github.com/Microsoft/TypeScript/wiki/Compiler-Options) for more details.

**Comment by jwbay** [comment]
I'm seeing this as well, and managed to get a minimal failing test case at https://github.com/jwbay/gulp-typescript/tree/sourcemap-casing/test/sourceMaps

I had a hard time reproducing on a smaller scale than my actual project, but from what I managed to find, this bug surfaces when you reference files by import in a directory higher than the target tsconfig.json. Definition files resolved via triple slash references don't seem to trigger it.

**Comment by jwbay** [comment]
Found the culprit, fix looks like this: https://github.com/jwbay/gulp-typescript/compare/windows-tests...jwbay:sourcemaps?expand=1

From what I could tell there's no reason to lowercase any paths in that block, but I just made the minimal change to fix the bug.

The fix branch (and its test baselines) are built off #294. I can follow up with a PR if/once that is merged, or someone with a mac/linux box can target something similar against master.
